File: Archiver/app_code/clsListener.cs
using System;
using System.Data;
using global::System.Data.SqlClient;
using Microsoft.VisualBasic;
using Microsoft.VisualBasic.CompilerServices;
using MODI;

namespace EcmArchiver
{
    public class clsListener : clsDatabaseARCH
    {
        private clsLogging LOG = new clsLogging();
        private clsDma DMA = new clsDma();
        private bool PauseOn = true;
        private bool PauseOff = false;

        // Static gActiveListeners As New LinkedList(Of String)

        public bool AddDirListener(string DirGuid, bool AdminDisabled, bool ListenerLoaded, bool ListenerActive, bool ListenerPaused, bool ListenDirectory, bool ListenSubDirectory, string MachineName)
        {
            var tGuid = new Guid(DirGuid);
            bool B = true;
            string S = "";
            int iCnt = iCount("Select count(*) from DirectoryListener where DirGuid = '" + DirGuid + "'");
            if (iCnt > 0)
            {
                S = "Update DirectoryListener set AdminDisabled = 0 where DirGuid = '" + DirGuid + "' ";
                B = ExecuteSqlNewConn(S, false);
                S = "Update DirectoryListener set ListenerLoaded = 0 where DirGuid = '" + DirGuid + "' ";
                B = ExecuteSqlNewConn(S, false);
                S = "Update DirectoryListener set ListenerActive = 0 where DirGuid = '" + DirGuid + "' ";
                B = ExecuteSqlNewConn(S, false);
                S = "Update DirectoryListener set ListenerPaused = 0 where DirGuid = '" + DirGuid + "' ";
                B = ExecuteSqlNewConn(S, false);
                if (ListenDirectory == true)
                {
                    S = "Update DirectoryListener set ListenDirectory = 1 where DirGuid = '" + DirGuid + "' ";
                    B = ExecuteSqlNewConn(S, false);
                }
                else
                {
                    S = "Update DirectoryListener set ListenDirectory = 0 where DirGuid = '" + DirGuid + "' ";
                    B = ExecuteSqlNewConn(S, false);
                }

                if (ListenSubDirectory == true)
                {
                    S = "Update DirectoryListener set ListenSubDirectory = 1 where DirGuid = '" + DirGuid + "' ";
                    B = ExecuteSqlNewConn(S, false);
                }
                else
                {
                    S = "Update DirectoryListener set ListenSubDirectory = 0 where DirGuid = '" + DirGuid + "' ";
                    B = ExecuteSqlNewConn(S, false);
                }
            }
            else
            {
                S = S + " INSERT INTO [DirectoryListener]" + Constants.vbCrLf;
                S = S + "            ([UserID]" + Constants.vbCrLf;
                S = S + "            ,[AdminDisabled]" + Constants.vbCrLf;
                S = S + "            ,[ListenerLoaded]" + Constants.vbCrLf;
                S = S + "            ,[ListenerActive]" + Constants.vbCrLf;
                S = S + "            ,[ListenerPaused]" + Constants.vbCrLf;
                S = S + "            ,[ListenDirectory]" + Constants.vbCrLf;
                S = S + "            ,[ListenSubDirectory]" + Constants.vbCrLf;
                S = S + "            ,[DirGuid]" + Constants.vbCrLf;
                S = S + "            ,[MachineName])";
                S = S + "      VALUES";
                S = S + "            ('" + modGlobals.gCurrUserGuidID + "'" + Constants.vbCrLf;
                S = S + "            ," + Conversion.Val(AdminDisabled) + Constants.vbCrLf;
                S = S + "            ," + Conversion.Val(ListenerLoaded) + Constants.vbCrLf;
                S = S + "            ," + Conversion.Val(ListenerActive) + Constants.vbCrLf;
                S = S + "            ," + Conversion.Val(ListenerPaused) + Constants.vbCrLf;
                S = S + "            ," + Conversion.Val(ListenDirectory) + Constants.vbCrLf;
                S = S + "            ," + Conversion.Val(ListenSubDirectory) + Constants.vbCrLf;
                S = S + "            ,'" + tGuid.ToString() + "'" + Constants.vbCrLf;
                S = S + "            ,'" + MachineName + "')";
                B = ExecuteSqlNewConn(S, false);
            }

            return B;
        }

        public bool deleteDirListener(string DirGuid)
        {
            bool B = true;
            string S = "";
            S = S + " delete from [DirectoryListener] where DirGuid = '" + DirGuid + "'";
            B = ExecuteSqlNewConn(S, false);
            return B;
        }

        public bool TurnOffSubDirListener(string DirGuid)
        {
            bool B = true;
            string S = "";
            S = S + " update [DirectoryListener] set ListenSubDirectory = 0 where DirGuid = '" + DirGuid + "' ";
            B = ExecuteSqlNewConn(S, false);
            if (B)
            {
                PauseDirListener(DirGuid, PauseOn);
            }

            return B;
        }

        public bool TurnOffListener(string DirGuid)
        {
            bool B = true;
            string S = "";
            S = S + " update [DirectoryListener] set ListenDirectory = 0 where DirGuid = '" + DirGuid + "' ";
            B = ExecuteSqlNewConn(S, false);
            if (B)
            {
                PauseDirListener(DirGuid, PauseOn);
            }

            return B;
        }

        public bool PauseDirListener(string DirGuid, bool bPause)
        {
            bool B = true;
            string S = "";
            if (bPause == true)
            {
                S = S + " update [DirectoryListener] set ListenerPaused = 1 where DirGuid = '" + DirGuid + "' ";
                int I = modGlobals.gActiveListeners.IndexOfKey(DirGuid);
                if (I >= 0)
                {
                    modGlobals.gActiveListeners[DirGuid] = PauseOn;
                }
            }
            else
            {
                S = S + " update [DirectoryListener] set ListenerPaused = 0 where DirGuid = '" + DirGuid + "' ";
                int I = modGlobals.gActiveListeners.IndexOfKey(DirGuid);
                if (I >= 0)
                {
                    modGlobals.gActiveListeners[DirGuid] = PauseOff;
                }
            }

            B = ExecuteSqlNewConn(S, false);
            return B;
        }

        public bool DisableDirListenerActive(string DirGuid)
        {
            bool B = true;
            string S = "";
            S = S + " update [DirectoryListener] set ListenerActive = 0, ListenerPaused = 1 where DirGuid = '" + DirGuid + "' ";
            B = ExecuteSqlNewConn(S, false);
            if (B)
            {
                int I = modGlobals.gActiveListeners.IndexOfKey(DirGuid);
                if (I >= 0)
                {
                    modGlobals.gActiveListeners.Values[I] = PauseOn;
                }
            }

            return B;
        }

        public bool EnableDirListenerActive(string DirGuid)
        {
            bool B = true;
            string S = "";
            S = S + " update [DirectoryListener] set ListenerActive = 1, ListenerPaused = 0 where DirGuid = '" + DirGuid + "' ";
            B = ExecuteSqlNewConn(S, false);
            if (B)
            {
                int I = modGlobals.gActiveListeners.IndexOfKey(DirGuid);
                if (I >= 0)
                {
                    modGlobals.gActiveListeners.Values[I] = PauseOff;
                }
            }

            return B;
        }

        public bool DisableDirListenerAdmin(string DirGuid)
        {
            bool B = true;
            string S = "";
            S = S + " update [DirectoryListener] set AdminDisabled = 1 where DirGuid = '" + DirGuid + "' ";
            B = ExecuteSqlNewConn(S, false);
            return B;
        }

        public bool EnableDirListenerAdmin(string DirGuid)
        {
            bool B = true;
            string S = "";
            S = S + " update [DirectoryListener] set AdminDisabled = 0 where DirGuid = '" + DirGuid + "' ";
            B = ExecuteSqlNewConn(S, false);
            return B;
        }

        public bool AddListenerFile(string DirGuid, string SourceFile, string MachineName)
        {
            bool B = true;
            string S = "";
            S = S + " INSERT INTO [DirectoryListenerFiles]";
            S = S + "            ([DirGuid]";
            S = S + "            ,[SourceFile]";
            S = S + "            ,[Archived]";
            S = S + "            ,[EntryDate]";
            S = S + "            ,[UserID]";
            S = S + "            ,[MachineName])";
            S = S + "      VALUES";
            S = S + "            ('" + DirGuid + "'";
            S = S + "            ,'" + SourceFile + "'";
            S = S + "            ,0";
            S = S + "            ,'" + DateAndTime.Now.ToString() + "'";
            S = S + "            ,'" + modGlobals.gCurrUserGuidID + "'";
            S = S + "            ,'" + MachineName + "'";
            B = ExecuteSqlNewConn(S, false);
            return B;
        }

        public bool setDirListernerON(string DirGuid)
        {
            bool B = true;
            string S = "";
            S = S + " Update Directory set ListenForChanges = 1, ListenDirectory = 1 where DirGuid = '" + DirGuid + "' ";
            B = ExecuteSqlNewConn(S, false);
            if (B)
            {
                PauseDirListener(DirGuid, PauseOff);
            }

            return B;
        }

        public bool setSubDirListernerON(string DirGuid)
        {
            bool B = true;
            string S = "";
            S = S + " Update Directory set ListenSubDirectory = 1 where DirGuid = '" + DirGuid + "' ";
            B = ExecuteSqlNewConn(S, false);
            if (B)
            {
                PauseDirListener(DirGuid, PauseOff);
            }

            return B;
        }

        public bool setDirListernerOFF(string DirGuid)
        {
            bool B = true;
            string S = "";
            S = S + " Update Directory set ListenForChanges = 0, ListenDirectory = 0, ListenSubDirectory = 0 where DirGuid = '" + DirGuid + "' ";
            B = ExecuteSqlNewConn(S, false);
            if (B)
            {
                PauseDirListener(DirGuid, PauseOn);
            }

            return B;
        }

        public int LoadListeners(string TgtMachineName)
        {
            int NbrListeners = 0;
            string S = "";
            S = S + " SELECT     DirectoryListener.UserID, DirectoryListener.AdminDisabled, DirectoryListener.ListenerLoaded, DirectoryListener.ListenerActive, DirectoryListener.ListenerPaused, ";
            S = S + " DirectoryListener.ListenDirectory, DirectoryListener.ListenSubDirectory, DirectoryListener.DirGuid, DirectoryListener.MachineName, Directory.FQN";
            S = S + " FROM         DirectoryListener INNER JOIN";
            S = S + " Directory ON DirectoryListener.DirGuid = Directory.DirGuid";
            S = S + " WHERE     (DirectoryListener.MachineName = '" + TgtMachineName + "') AND (DirectoryListener.ListenerActive = 1)";
            string UserID = "";
            bool AdminDisabled = false;
            bool ListenerLoaded = false;
            bool ListenerActive = false;
            bool ListenerPaused = false;
            bool ListenDirectory = false;
            bool ListenSubDirectory = false;
            string DirGuid = "";
            string MachineName = "";
            string FQN = "";
            SqlDataReader RSData = null;
            string CS = getRepoConnStr();
            var CONN = new SqlConnection(CS);
            CONN.Open();
            var command = new SqlCommand(S, CONN);
            RSData = command.ExecuteReader();
            if (RSData.HasRows)
            {
                while (RSData.Read())
                {
                    NbrListeners += 1;
                    UserID = RSData.GetValue(0).ToString();
                    AdminDisabled = Conversions.ToBoolean(RSData.GetValue(1).ToString());
                    ListenerLoaded = RSData.GetBoolean(2);
                    ListenerActive = RSData.GetBoolean(3);
                    ListenerPaused = RSData.GetBoolean(4);
                    ListenDirectory = RSData.GetBoolean(5);
                    ListenSubDirectory = RSData.GetBoolean(6);
                    DirGuid = RSData.GetValue(7).ToString();
                    MachineName = RSData.GetValue(8).ToString();
                    FQN = RSData.GetValue(9).ToString();
                    if (modGlobals.gActiveListeners.IndexOfKey(DirGuid) >= 0)
                    {
                    }
                    else
                    {
                        var DL = new clsDirListener();
                        DL.WatchDirectory = FQN;
                        DL.DirGuid = DirGuid;
                        DL.Machinename = MachineName;
                        DL.StartListening(ListenSubDirectory);
                        try
                        {
                            modGlobals.gActiveListeners.Add(DirGuid, ListenerPaused);
                        }
                        catch (Exception ex)
                        {
                            Console.WriteLine(ex.Message);
                        }
                    }
                }
            }
            else
            {
                LOG.WriteToArchiveLog("Notification: No directory listeners found for machine " + TgtMachineName + ".");
            }

            if (!RSData.IsClosed)
            {
                RSData.Close();
            }

            RSData = null;
            command.Dispose();
            command = null;
            if (CONN.State == ConnectionState.Open)
            {
                CONN.Close();
            }

            CONN.Dispose();
            return NbrListeners;
        }

        public void PauseListeners(string TgtMachineName, bool bPause)
        {
            string S = "";
            if (bPause == true)
            {
                S = S + " Update [DirectoryListener] set ListenerPaused = 1 ";
            }
            else
            {
                S = S + " Update [DirectoryListener] set ListenerPaused = 0 ";
            }

            S = S + " WHERE  DirectoryListener.MachineName = '" + TgtMachineName + "'";
            bool B = ExecuteSqlNewConn(90400, S);
            for (int I = 0, loopTo = modGlobals.gActiveListeners.Count - 1; I <= loopTo; I++)
            {
                S = modGlobals.gActiveListeners.Keys[I].ToString();
                modGlobals.gActiveListeners[S] = bPause;
            }
        }
    }
}

File: Archiver/app_code/clsListener.vb
Imports System.Data.SqlClient
Imports System.Collections
Imports System.IO
Imports System.Data.Sql

Public Class clsListener
    Inherits clsDatabaseARCH

    Dim LOG As New clsLogging
    Dim DMA As New clsDma

    Dim PauseOn As Boolean = True
    Dim PauseOff As Boolean = False

    'Static gActiveListeners As New LinkedList(Of String)

    Function AddDirListener(ByVal DirGuid As String,
                            ByVal AdminDisabled As Boolean,
                            ByVal ListenerLoaded As Boolean,
                            ByVal ListenerActive As Boolean,
                            ByVal ListenerPaused As Boolean,
                            ByVal ListenDirectory As Boolean,
                            ByVal ListenSubDirectory As Boolean,
                            ByVal MachineName As String) As Boolean

        Dim tGuid As Guid = New Guid(DirGuid)
        Dim B As Boolean = True
        Dim S As String = ""
        Dim iCnt As Integer = iCount("Select count(*) from DirectoryListener where DirGuid = '" + DirGuid + "'")

        If iCnt > 0 Then
            S = "Update DirectoryListener set AdminDisabled = 0 where DirGuid = '" + DirGuid + "' "
            B = ExecuteSqlNewConn(S, False)

            S = "Update DirectoryListener set ListenerLoaded = 0 where DirGuid = '" + DirGuid + "' "
            B = ExecuteSqlNewConn(S, False)

            S = "Update DirectoryListener set ListenerActive = 0 where DirGuid = '" + DirGuid + "' "
            B = ExecuteSqlNewConn(S, False)

            S = "Update DirectoryListener set ListenerPaused = 0 where DirGuid = '" + DirGuid + "' "
            B = ExecuteSqlNewConn(S, False)

            If ListenDirectory = True Then
                S = "Update DirectoryListener set ListenDirectory = 1 where DirGuid = '" + DirGuid + "' "
                B = ExecuteSqlNewConn(S, False)
            Else
                S = "Update DirectoryListener set ListenDirectory = 0 where DirGuid = '" + DirGuid + "' "
                B = ExecuteSqlNewConn(S, False)
            End If
            If ListenSubDirectory = True Then
                S = "Update DirectoryListener set ListenSubDirectory = 1 where DirGuid = '" + DirGuid + "' "
                B = ExecuteSqlNewConn(S, False)
            Else
                S = "Update DirectoryListener set ListenSubDirectory = 0 where DirGuid = '" + DirGuid + "' "
                B = ExecuteSqlNewConn(S, False)
            End If

        Else
            S = S + " INSERT INTO [DirectoryListener]" + vbCrLf
            S = S + "            ([UserID]" + vbCrLf
            S = S + "            ,[AdminDisabled]" + vbCrLf
            S = S + "            ,[ListenerLoaded]" + vbCrLf
            S = S + "            ,[ListenerActive]" + vbCrLf
            S = S + "            ,[ListenerPaused]" + vbCrLf
            S = S + "            ,[ListenDirectory]" + vbCrLf
            S = S + "            ,[ListenSubDirectory]" + vbCrLf
            S = S + "            ,[DirGuid]" + vbCrLf
            S = S + "            ,[MachineName])"
            S = S + "      VALUES"
            S = S + "            ('" + gCurrUserGuidID + "'" + vbCrLf
            S = S + "            ," & Val(AdminDisabled) & vbCrLf
            S = S + "            ," & Val(ListenerLoaded) & vbCrLf
            S = S + "            ," & Val(ListenerActive) & vbCrLf
            S = S + "            ," & Val(ListenerPaused) & vbCrLf
            S = S + "            ," & Val(ListenDirectory) & vbCrLf
            S = S + "            ," & Val(ListenSubDirectory) & vbCrLf
            S = S + "            ,'" & tGuid.ToString & "'" + vbCrLf
            S = S + "            ,'" + MachineName + "')"

            B = ExecuteSqlNewConn(S, False)
        End If



        Return B
    End Function

    Function deleteDirListener(ByVal DirGuid As String) As Boolean

        Dim B As Boolean = True
        Dim S As String = ""
        S = S + " delete from [DirectoryListener] where DirGuid = '" + DirGuid + "'"
        B = ExecuteSqlNewConn(S, False)

        Return B
    End Function

    Function TurnOffSubDirListener(ByVal DirGuid As String) As Boolean

        Dim B As Boolean = True
        Dim S As String = ""

        S = S + " update [DirectoryListener] set ListenSubDirectory = 0 where DirGuid = '" + DirGuid + "' "

        B = ExecuteSqlNewConn(S, False)

        If B Then
            PauseDirListener(DirGuid, PauseOn)
        End If

        Return B
    End Function

    Function TurnOffListener(ByVal DirGuid As String) As Boolean

        Dim B As Boolean = True
        Dim S As String = ""

        S = S + " update [DirectoryListener] set ListenDirectory = 0 where DirGuid = '" + DirGuid + "' "

        B = ExecuteSqlNewConn(S, False)

        If B Then
            PauseDirListener(DirGuid, PauseOn)
        End If

        Return B
    End Function

    Function PauseDirListener(ByVal DirGuid As String, ByVal bPause As Boolean) As Boolean

        Dim B As Boolean = True
        Dim S As String = ""
        If bPause = True Then
            S = S + " update [DirectoryListener] set ListenerPaused = 1 where DirGuid = '" + DirGuid + "' "
            Dim I As Integer = gActiveListeners.IndexOfKey(DirGuid)
            If I >= 0 Then
                gActiveListeners.Item(DirGuid) = PauseOn
            End If
        Else
            S = S + " update [DirectoryListener] set ListenerPaused = 0 where DirGuid = '" + DirGuid + "' "
            Dim I As Integer = gActiveListeners.IndexOfKey(DirGuid)
            If I >= 0 Then
                gActiveListeners.Item(DirGuid) = PauseOff
            End If
        End If

        B = ExecuteSqlNewConn(S, False)

        Return B
    End Function

    Function DisableDirListenerActive(ByVal DirGuid As String) As Boolean

        Dim B As Boolean = True
        Dim S As String = ""

        S = S + " update [DirectoryListener] set ListenerActive = 0, ListenerPaused = 1 where DirGuid = '" + DirGuid + "' "

        B = ExecuteSqlNewConn(S, False)

        If B Then
            Dim I As Integer = gActiveListeners.IndexOfKey(DirGuid)
            If I >= 0 Then
                gActiveListeners.Values(I) = PauseOn
            End If
        End If

        Return B
    End Function
    Function EnableDirListenerActive(ByVal DirGuid As String) As Boolean

        Dim B As Boolean = True
        Dim S As String = ""

        S = S + " update [DirectoryListener] set ListenerActive = 1, ListenerPaused = 0 where DirGuid = '" + DirGuid + "' "

        B = ExecuteSqlNewConn(S, False)

        If B Then
            Dim I As Integer = gActiveListeners.IndexOfKey(DirGuid)
            If I >= 0 Then
                gActiveListeners.Values(I) = PauseOff
            End If
        End If

        Return B
    End Function

    Function DisableDirListenerAdmin(ByVal DirGuid As String) As Boolean

        Dim B As Boolean = True
        Dim S As String = ""

        S = S + " update [DirectoryListener] set AdminDisabled = 1 where DirGuid = '" + DirGuid + "' "

        B = ExecuteSqlNewConn(S, False)

        Return B
    End Function

    Function EnableDirListenerAdmin(ByVal DirGuid As String) As Boolean

        Dim B As Boolean = True
        Dim S As String = ""

        S = S + " update [DirectoryListener] set AdminDisabled = 0 where DirGuid = '" + DirGuid + "' "

        B = ExecuteSqlNewConn(S, False)

        Return B
    End Function

    Function AddListenerFile(ByVal DirGuid As String, ByVal SourceFile As String, ByVal MachineName As String) As Boolean
        Dim B As Boolean = True
        Dim S As String = ""
        S = S + " INSERT INTO [DirectoryListenerFiles]"
        S = S + "            ([DirGuid]"
        S = S + "            ,[SourceFile]"
        S = S + "            ,[Archived]"
        S = S + "            ,[EntryDate]"
        S = S + "            ,[UserID]"
        S = S + "            ,[MachineName])"
        S = S + "      VALUES"
        S = S + "            ('" + DirGuid + "'"
        S = S + "            ,'" + SourceFile + "'"
        S = S + "            ,0"
        S = S + "            ,'" + Now.ToString + "'"
        S = S + "            ,'" + gCurrUserGuidID + "'"
        S = S + "            ,'" + MachineName + "'"

        B = ExecuteSqlNewConn(S, False)

        Return B
    End Function

    Function setDirListernerON(ByVal DirGuid As String) As Boolean
        Dim B As Boolean = True
        Dim S As String = ""
        S = S + " Update Directory set ListenForChanges = 1, ListenDirectory = 1 where DirGuid = '" + DirGuid + "' "
        B = ExecuteSqlNewConn(S, False)
        If B Then
            PauseDirListener(DirGuid, PauseOff)
        End If
        Return B
    End Function

    Function setSubDirListernerON(ByVal DirGuid As String) As Boolean
        Dim B As Boolean = True
        Dim S As String = ""
        S = S + " Update Directory set ListenSubDirectory = 1 where DirGuid = '" + DirGuid + "' "
        B = ExecuteSqlNewConn(S, False)
        If B Then
            PauseDirListener(DirGuid, PauseOff)
        End If
        Return B
    End Function

    Function setDirListernerOFF(ByVal DirGuid As String) As Boolean
        Dim B As Boolean = True
        Dim S As String = ""
        S = S + " Update Directory set ListenForChanges = 0, ListenDirectory = 0, ListenSubDirectory = 0 where DirGuid = '" + DirGuid + "' "
        B = ExecuteSqlNewConn(S, False)

        If B Then
            PauseDirListener(DirGuid, PauseOn)
        End If

        Return B
    End Function

    Function LoadListeners(ByVal TgtMachineName As String) As Integer

        Dim NbrListeners As Integer = 0

        Dim S As String = ""
        S = S + " SELECT     DirectoryListener.UserID, DirectoryListener.AdminDisabled, DirectoryListener.ListenerLoaded, DirectoryListener.ListenerActive, DirectoryListener.ListenerPaused, "
        S = S + " DirectoryListener.ListenDirectory, DirectoryListener.ListenSubDirectory, DirectoryListener.DirGuid, DirectoryListener.MachineName, Directory.FQN"
        S = S + " FROM         DirectoryListener INNER JOIN"
        S = S + " Directory ON DirectoryListener.DirGuid = Directory.DirGuid"
        S = S + " WHERE     (DirectoryListener.MachineName = '" + TgtMachineName + "') AND (DirectoryListener.ListenerActive = 1)"

        Dim UserID As String = ""
        Dim AdminDisabled As Boolean = False
        Dim ListenerLoaded As Boolean = False
        Dim ListenerActive As Boolean = False
        Dim ListenerPaused As Boolean = False
        Dim ListenDirectory As Boolean = False
        Dim ListenSubDirectory As Boolean = False
        Dim DirGuid As String = ""
        Dim MachineName As String = ""
        Dim FQN As String = ""

        Dim RSData As SqlDataReader = Nothing
        Dim CS As String = getRepoConnStr()
        Dim CONN As New SqlConnection(CS)
        CONN.Open()
        Dim command As New SqlCommand(S, CONN)
        RSData = command.ExecuteReader()
        If RSData.HasRows Then
            Do While RSData.Read()
                NbrListeners += 1
                UserID = RSData.GetValue(0).ToString
                AdminDisabled = RSData.GetValue(1).ToString
                ListenerLoaded = RSData.GetBoolean(2)
                ListenerActive = RSData.GetBoolean(3)
                ListenerPaused = RSData.GetBoolean(4)
                ListenDirectory = RSData.GetBoolean(5)
                ListenSubDirectory = RSData.GetBoolean(6)
                DirGuid = RSData.GetValue(7).ToString
                MachineName = RSData.GetValue(8).ToString
                FQN = RSData.GetValue(9).ToString

                If gActiveListeners.IndexOfKey(DirGuid) >= 0 Then
                Else
                    Dim DL As New clsDirListener
                    DL.WatchDirectory = FQN
                    DL.DirGuid = DirGuid
                    DL.Machinename = MachineName
                    DL.StartListening(ListenSubDirectory)
                    Try
                        gActiveListeners.Add(DirGuid, ListenerPaused)
                    Catch ex As Exception
                        Console.WriteLine(ex.Message)
                    End Try

                End If

            Loop
        Else
            LOG.WriteToArchiveLog("Notification: No directory listeners found for machine " + TgtMachineName + ".")
        End If

        If Not RSData.IsClosed Then
            RSData.Close()
        End If
        RSData = Nothing
        command.Dispose()
        command = Nothing

        If CONN.State = ConnectionState.Open Then
            CONN.Close()
        End If
        CONN.Dispose()
        Return NbrListeners
    End Function

    Sub PauseListeners(ByVal TgtMachineName As String, ByVal bPause As Boolean)

        Dim S As String = ""
        If bPause = True Then
            S = S + " Update [DirectoryListener] set ListenerPaused = 1 "
        Else
            S = S + " Update [DirectoryListener] set ListenerPaused = 0 "
        End If
        S = S + " WHERE  DirectoryListener.MachineName = '" + TgtMachineName + "'"

        Dim B As Boolean = ExecuteSqlNewConn(90400, S)

        For I As Integer = 0 To gActiveListeners.Count - 1
            S = gActiveListeners.Keys(I).ToString
            gActiveListeners.Item(S) = bPause
        Next

    End Sub

End Class
